File: services/optimizer/routing/pack3d.py
"""3D extreme-point packing; OR-Tools CP-SAT optional for tiny instances."""

from __future__ import annotations

from typing import Any


VEHICLES = {
    "407": {"l": 2.8, "w": 1.7, "h": 1.8, "maxWeightKg": 2500, "vehicle_type": "407"},
    "14ft": {"l": 4.2, "w": 2.1, "h": 2.0, "maxWeightKg": 4000, "vehicle_type": "14ft"},
    "20ft": {"l": 5.9, "w": 2.35, "h": 2.3, "maxWeightKg": 7000, "vehicle_type": "20ft"},
    "32ft": {"l": 9.5, "w": 2.4, "h": 2.5, "maxWeightKg": 12000, "vehicle_type": "32ft"},
}


def _expand(boxes: list[dict[str, Any]]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for b in boxes:
        qty = max(1, min(int(b.get("qty") or 1), 40))
        for i in range(qty):
            out.append(
                {
                    "id": f"{b.get('id')}#{i + 1}" if qty > 1 else str(b.get("id")),
                    "l": float(b.get("l") or 0.3),
                    "w": float(b.get("w") or 0.3),
                    "h": float(b.get("h") or 0.3),
                    "weight_kg": float(b.get("weight_kg") or 5),
                }
            )
    out.sort(key=lambda x: -(x["l"] * x["w"] * x["h"]))
    return out


def _overlaps(a: dict[str, Any], b: dict[str, Any]) -> bool:
    return not (
        a["x"] + a["l"] <= b["x"]
        or b["x"] + b["l"] <= a["x"]
        or a["y"] + a["w"] <= b["y"]
        or b["y"] + b["w"] <= a["y"]
        or a["z"] + a["h"] <= b["z"]
        or b["z"] + b["h"] <= a["z"]
    )


def pack_3d(boxes: list[dict[str, Any]], vehicle_type: str = "14ft", max_loads: int = 12) -> dict[str, Any]:
    container = VEHICLES.get(vehicle_type, VEHICLES["14ft"])
    remaining = _expand(boxes)
    loads: list[dict[str, Any]] = []
    unpacked: list[str] = []

    while remaining and len(loads) < max_loads:
        placed: list[dict[str, Any]] = []
        weight = 0.0
        candidates = [{"x": 0.0, "y": 0.0, "z": 0.0}]
        still: list[dict[str, Any]] = []

        for box in remaining:
            if weight + box["weight_kg"] > container["maxWeightKg"]:
                still.append(box)
                continue
            placed_one = None
            orients = [
                (box["l"], box["w"], box["h"]),
                (box["w"], box["l"], box["h"]),
                (box["l"], box["h"], box["w"]),
            ]
            for pt in candidates:
                for l, w, h in orients:
                    cand = {
                        "id": box["id"],
                        "x": pt["x"],
                        "y": pt["y"],
                        "z": pt["z"],
                        "l": l,
                        "w": w,
                        "h": h,
                        "weight_kg": box["weight_kg"],
                    }
                    if (
                        cand["x"] + l > container["l"] + 1e-9
                        or cand["y"] + w > container["w"] + 1e-9
                        or cand["z"] + h > container["h"] + 1e-9
                    ):
                        continue
                    if any(_overlaps(cand, p) for p in placed):
                        continue
                    placed_one = cand
                    break
                if placed_one:
                    break
            if not placed_one:
                still.append(box)
                continue
            placed.append(placed_one)
            weight += box["weight_kg"]
            candidates.extend(
                [
                    {"x": placed_one["x"] + placed_one["l"], "y": placed_one["y"], "z": placed_one["z"]},
                    {"x": placed_one["x"], "y": placed_one["y"] + placed_one["w"], "z": placed_one["z"]},
                    {"x": placed_one["x"], "y": placed_one["y"], "z": placed_one["z"] + placed_one["h"]},
                ]
            )

        if not placed:
            unpacked.extend([b["id"] for b in still + remaining])
            break

        used_vol = sum(p["l"] * p["w"] * p["h"] for p in placed)
        cap_vol = container["l"] * container["w"] * container["h"]
        fill_vol = round(100.0 * used_vol / cap_vol, 1)
        fill_wt = round(100.0 * weight / container["maxWeightKg"], 1)
        loads.append(
            {
                "load_id": f"3D-{len(loads) + 1:03d}",
                "vehicle_type": container["vehicle_type"],
                "placements": placed,
                "used_volume_m3": round(used_vol, 3),
                "used_weight_kg": round(weight, 1),
                "fill_volume_pct": fill_vol,
                "fill_weight_pct": fill_wt,
                "fill_pct": max(fill_vol, fill_wt),
            }
        )
        remaining = still

    if remaining and len(loads) >= max_loads:
        unpacked.extend(b["id"] for b in remaining)

    avg = round(sum(l["fill_pct"] for l in loads) / len(loads), 1) if loads else 0
    return {
        "engine": "extreme-point-3d",
        "capacity": container,
        "loads": loads,
        "unpacked": list(dict.fromkeys(unpacked)),
        "summary": {
            "loads": len(loads),
            "avgFillPct": avg,
            "boxesPacked": sum(len(l["placements"]) for l in loads),
            "boxesUnpacked": len(unpacked),
        },
    }

File: apps/web/src/lib/math/pack3d.ts
/**
 * Extreme-point 3D bin packing into a vehicle container.
 * Places boxes (L×W×H) without overlap; respects weight.
 * Inspired by pallet-loading / 3D bin packing skills.
 */

export interface Box3D {
  id: string;
  l: number; // m
  w: number;
  h: number;
  weight_kg: number;
  qty?: number;
}

export interface Container3D {
  l: number;
  w: number;
  h: number;
  maxWeightKg: number;
  vehicle_type: string;
}

export interface PlacedBox {
  id: string;
  x: number;
  y: number;
  z: number;
  l: number;
  w: number;
  h: number;
  weight_kg: number;
}

export interface PackedLoad {
  load_id: string;
  vehicle_type: string;
  placements: PlacedBox[];
  used_volume_m3: number;
  used_weight_kg: number;
  fill_volume_pct: number;
  fill_weight_pct: number;
  fill_pct: number;
}

const VEHICLE_3D: Record<string, Container3D> = {
  "407": { l: 2.8, w: 1.7, h: 1.8, maxWeightKg: 2500, vehicle_type: "407" },
  "14ft": { l: 4.2, w: 2.1, h: 2.0, maxWeightKg: 4000, vehicle_type: "14ft" },
  "20ft": { l: 5.9, w: 2.35, h: 2.3, maxWeightKg: 7000, vehicle_type: "20ft" },
  "32ft": { l: 9.5, w: 2.4, h: 2.5, maxWeightKg: 12000, vehicle_type: "32ft" },
};

export function vehicleContainer(type: string): Container3D {
  return VEHICLE_3D[type] ?? VEHICLE_3D["14ft"]!;
}

function overlaps(a: PlacedBox, b: PlacedBox): boolean {
  return !(
    a.x + a.l <= b.x ||
    b.x + b.l <= a.x ||
    a.y + a.w <= b.y ||
    b.y + b.w <= a.y ||
    a.z + a.h <= b.z ||
    b.z + b.h <= a.z
  );
}

function fits(box: PlacedBox, c: Container3D): boolean {
  return box.x + box.l <= c.l + 1e-9 && box.y + box.w <= c.w + 1e-9 && box.z + box.h <= c.h + 1e-9;
}

/** Expand qty into individual boxes; sort by volume descending (FFD). */
function expandBoxes(boxes: Box3D[]): Box3D[] {
  const out: Box3D[] = [];
  for (const b of boxes) {
    const q = Math.max(1, Math.min(b.qty ?? 1, 40));
    for (let i = 0; i < q; i++) {
      out.push({
        id: q > 1 ? `${b.id}#${i + 1}` : b.id,
        l: b.l,
        w: b.w,
        h: b.h,
        weight_kg: b.weight_kg,
      });
    }
  }
  return out.sort((a, b) => b.l * b.w * b.h - a.l * a.w * a.h);
}

function tryPlace(
  box: Box3D,
  container: Container3D,
  placed: PlacedBox[],
  candidates: { x: number; y: number; z: number }[],
): PlacedBox | null {
  const orients: [number, number, number][] = [
    [box.l, box.w, box.h],
    [box.l, box.h, box.w],
    [box.w, box.l, box.h],
    [box.w, box.h, box.l],
    [box.h, box.l, box.w],
    [box.h, box.w, box.l],
  ];

  for (const pt of candidates) {
    for (const [l, w, h] of orients) {
      const cand: PlacedBox = {
        id: box.id,
        x: pt.x,
        y: pt.y,
        z: pt.z,
        l,
        w,
        h,
        weight_kg: box.weight_kg,
      };
      if (!fits(cand, container)) continue;
      if (placed.some((p) => overlaps(cand, p))) continue;
      return cand;
    }
  }
  return null;
}

export function pack3dIntoVehicles(
  boxes: Box3D[],
  vehicleType = "14ft",
  maxLoads = 12,
): {
  engine: "extreme-point-3d";
  capacity: Container3D;
  loads: PackedLoad[];
  unpacked: string[];
  summary: {
    loads: number;
    avgFillPct: number;
    boxesPacked: number;
    boxesUnpacked: number;
  };
} {
  const items = expandBoxes(boxes);
  const container = vehicleContainer(vehicleType);
  const remaining = [...items];
  const loads: PackedLoad[] = [];
  const unpacked: string[] = [];

  while (remaining.length > 0 && loads.length < maxLoads) {
    const placed: PlacedBox[] = [];
    let weight = 0;
    const candidates: { x: number; y: number; z: number }[] = [{ x: 0, y: 0, z: 0 }];
    const still: Box3D[] = [];

    for (const box of remaining) {
      if (weight + box.weight_kg > container.maxWeightKg) {
        still.push(box);
        continue;
      }
      const pos = tryPlace(box, container, placed, candidates);
      if (!pos) {
        still.push(box);
        continue;
      }
      placed.push(pos);
      weight += box.weight_kg;
      candidates.push(
        { x: pos.x + pos.l, y: pos.y, z: pos.z },
        { x: pos.x, y: pos.y + pos.w, z: pos.z },
        { x: pos.x, y: pos.y, z: pos.z + pos.h },
      );
    }

    if (placed.length === 0) {
      // Cannot place any remaining — mark unpacked and stop
      unpacked.push(...still.map((b) => b.id), ...remaining.map((b) => b.id));
      break;
    }

    const usedVol = placed.reduce((s, p) => s + p.l * p.w * p.h, 0);
    const capVol = container.l * container.w * container.h;
    const fillVol = Math.round((usedVol / capVol) * 1000) / 10;
    const fillWt = Math.round((weight / container.maxWeightKg) * 1000) / 10;

    loads.push({
      load_id: `3D-${String(loads.length + 1).padStart(3, "0")}`,
      vehicle_type: container.vehicle_type,
      placements: placed,
      used_volume_m3: Math.round(usedVol * 1000) / 1000,
      used_weight_kg: Math.round(weight * 10) / 10,
      fill_volume_pct: fillVol,
      fill_weight_pct: fillWt,
      fill_pct: Math.max(fillVol, fillWt),
    });

    remaining.length = 0;
    remaining.push(...still);
  }

  if (remaining.length && loads.length >= maxLoads) {
    unpacked.push(...remaining.map((b) => b.id));
  }

  const avgFill =
    loads.length > 0
      ? Math.round((loads.reduce((s, l) => s + l.fill_pct, 0) / loads.length) * 10) / 10
      : 0;
  const boxesPacked = loads.reduce((s, l) => s + l.placements.length, 0);

  return {
    engine: "extreme-point-3d",
    capacity: container,
    loads,
    unpacked: Array.from(new Set(unpacked)),
    summary: {
      loads: loads.length,
      avgFillPct: avgFill,
      boxesPacked,
      boxesUnpacked: unpacked.length,
    },
  };
}

/** Derive boxes from order lines + sku dims (fallback defaults). */
export function boxesFromOrders(
  orders: Record<string, string | number | undefined>[],
  skus: Record<string, string | number | undefined>[],
): Box3D[] {
  const skuMap = new Map(skus.map((s) => [String(s.sku_id), s]));
  return orders
    .filter((o) => String(o.status ?? "open").toLowerCase() === "open")
    .slice(0, 30)
    .map((o) => {
      const sku = skuMap.get(String(o.sku_id));
      const qty = Number(o.qty) || 1;
      const vol = Number(sku?.volume_m3) || 0.02;
      // Approximate cube root for missing LWH
      const side = Math.cbrt(vol);
      return {
        id: String(o.order_id ?? o.sku_id),
        l: Number(sku?.length_m) || side,
        w: Number(sku?.width_m) || side,
        h: Number(sku?.height_m) || side,
        weight_kg: (Number(sku?.weight_kg) || 5) * Math.min(qty, 1),
        qty: Math.min(qty, 8),
      };
    });
}
